src/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -1433,6 +1433,79 @@
+        (first (jdbc/query-to-vec "select current_database(), current_user;"))
+        kill-blocking-qs (format
+                          "select pg_cancel_backend(l.pid)
+                            from pg_locks l inner join pg_stat_activity a on l.pid = a.pid
+                            where (a.datname = '%s' AND a.usename = '%s')
+                             and not (l.pid = pg_backend_pid() OR l.pid = %s)
+                             and l.granted = true
+                             and exists (select 1
+                                          from pg_locks
+                                          where pid = %s and mode = 'AccessExclusiveLock'
+                                           and granted = false)
+                            group by l.pid;"
+                          current_database current_user gc-pid gc-pid)]
+    (Thread.
+     #(with-noisy-failure
+        (try
+          ;; you can't use *db* value from the gc thread because the connection
+          ;; it has may already be blocked attempting to drop the partition.
+          ;; Binding *db* to the passed in db var will cause the killer thread
+          ;; to get a new Hikari connection
+          (binding [jdbc/*db* db]
+            ;; Q: is there any possible way that all queries could be in a blocked
+            ;; state where we're unable to get another connection from the pool
+            ;; to issue the kill command?
+            (while (not @gc-finished?)
+              (jdbc/do-commands kill-blocking-qs)
+              (Thread/sleep 1000)))
+          ;; Q: do we need to worry about other exceptions here as well?
+          (catch InterruptedException ex
+            ;; the killer thread may be interrupted by the GC thread if it's
+            ;; in the sleep or do-commands part of the while loop
+            true))))))
+
+(def gc-query-killer-timeout-ms
+  (env-config-for-db-ulong "PDB_GC_QUERY_KILLER_TIMEOUT_MS"
+                           (* 20 60 1000)))
+
+(defn maybe-launch-query-killer [drop-one candidate db]
+  (if (pos? gc-query-killer-timeout-ms)
+    (let [gc-pid (-> "select pg_backend_pid();"
+                     jdbc/query-to-vec
+                     first
+                     :pg_backend_pid)
+          gc-finished? (atom false)
+          query-killer (spawn-query-killer db gc-pid gc-finished?)]
+      (try
+        (.start query-killer)
+        (drop-one candidate)
+        (reset! gc-finished? true)
+        (finally
+          ;; make sure the GC killer thread has broken out of its loop
+          (.interrupt query-killer) ; Q: do we need an interrupt as well as the gc-finished? atom?
+          (.join query-killer gc-query-killer-timeout-ms)
+          (when (.isAlive query-killer)
+            ;; if the join above timed out and the killer thread is still alive
+            ;; log an ERROR because it could indicate a memory leak
+            (log/error "Unable to clean up the gc killer thread.
```

Comment:
It will only kill queries when it detects that the gc pid it was started with is blocked on an AccessExclusiveLock. It's possible if the killer thread got wedged and the same pid was used and blocked on that lock at a later point that it could wake up and kill queries it shouldn't. We could request a shutdown in this case if we wanted to be extra careful to never allow this to happen.

Reply:
I think since it has to be the same pid and an AccessExclusiveLock just logging an error should be ok. I don't know of anything else that would grab an exclusive lock like that except pgrepack and it'd do the same thing.